learning_wavelets/models/exact_recon_unet.py

```diff
@@ -0,0 +1,169 @@
+        self.n_layers = len(self.layers_n_channels)
+        self.layers_n_non_lins = layers_n_non_lins
+        self.non_linearity = non_linearity
+        self.down_convs = [
+            ConvBlock(
+                n_filters=n_channels,
+                kernel_size=self.kernel_size,
+                non_linearity=self.non_linearity,
+                n_non_lins=self.layers_n_non_lins,
+            ) for n_channels in self.layers_n_channels[:-1]
+        ]
+        self.down = MaxPooling2D(pool_size=(2, 2), padding='same')
+        self.bottom_conv = ConvBlock(
+            n_filters=self.layers_n_channels[-1],
+            kernel_size=self.kernel_size,
+            non_linearity=self.non_linearity,
+            n_non_lins=self.layers_n_non_lins,
+        )
+        self.up_convs = [
+            ConvBlock(
+                n_filters=n_channels,
+                kernel_size=self.kernel_size,
+                non_linearity=self.non_linearity,
+                n_non_lins=self.layers_n_non_lins,
+            ) for n_channels in self.layers_n_channels[:-1]
+        ]
+        self.ups = [
+            UpConv(
+                n_filters=n_channels,
+                kernel_size=self.kernel_size,
+            ) for n_channels in self.layers_n_channels[:-1]
+        ]
+        self.final_conv = Conv2D(
+            filters=self.n_output_channels,
+            kernel_size=1,
+            padding='same',
+            activation=None,
+        )
+
+    
+    def call(self, inputs):
+        scales = []
+        noisy_image = inputs[0]
+        noise_std = inputs[1]
+        outputs = noisy_image
+        outputs, crop_w, crop_h, w, h = pad_power_of_two(outputs, self.layers_n_channels)
+        for conv in self.down_convs:
+            outputs = conv(outputs)
+            scales.append(outputs)
+            outputs = self.down(outputs)
+        outputs = self.bottom_conv(outputs)
+        for scale, conv, up in zip(scales[::-1], self.up_convs[::-1], self.ups[::-1]):
+            outputs = up(outputs)
+            outputs = tf.concat([outputs, scale], axis=-1)
+            outputs = conv(outputs)
+        outputs = self.final_conv(outputs)
+        noise_std = tf.reshape(noise_std, shape=[tf.shape(noise_std)[0], 1, 1, 1])
+        outputs = tf.image.crop_to_bounding_box(outputs, int(floor(crop_h)), int(floor(crop_w)), h, w)
+        outputs = noisy_image - noise_std * outputs
+        return outputs
+    
+
+def pad_power_of_two(x, layers):
+    
+    diff = 2**(len(layers))
+
+    y = tf.convert_to_tensor(x).numpy()
+    h = y.shape[1]
+    w = y.shape[2]
+    
+    if(w % diff != 0):
+        pad_value_w = diff - w % diff 
+    else:
+        pad_value_w = 0
+    if(h % diff != 0):
+        pad_value_h = diff - h % diff 
+    else:
+        pad_value_h = 0
+    
+    
+    padding = tf.constant([(0,0), (int(floor(pad_value_h/2)), int(ceil(pad_value_h/2))), (int(floor(pad_value_w/2)), int(ceil(pad_value_w/2))), (0, 0)])
```

Comment:
be careful, the padding is not always going to be symmetric.

Typically, in bsd68, the images have odd shapes.

I also think you should not be using `tf.pad` but rather `https://www.tensorflow.org/api_docs/python/tf/image/resize_with_crop_or_pad` and this way you just need to keep track of the original image shape rather than the padding values (i.e. you can also avoid `crop_to_bounding_box`).